### PR TITLE
initial check in for caclient

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/istio/istio/issues/5903

### CODE REVIEW

File: security/cmd/node_agent_k8s/main.go [hunk 1]
```diff
@@ -80,7 +81,7 @@ func main() {
 type mockCAClient struct {
 }
 
-func (c *mockCAClient) CSRSign(csrPEM []byte, /*PEM-encoded certificate request*/
+func (c *mockCAClient) CSRSign(ctx context.Context, csrPEM []byte, /*PEM-encoded certificate request*/
```
Comment (wattli) [inline security/cmd/node_agent_k8s/main.go]: Why we need this "/*PEM-encoded certificate request*/", i don't think it's a proper comment for subjectID.

File: security/pkg/nodeagent/caClient/client.go [hunk 2]
```diff
@@ -14,7 +14,61 @@
...
+	"google.golang.org/grpc/metadata"
+
+	"istio.io/istio/pkg/log"
+	capb "istio.io/istio/security/proto/ca/v1alpha1"
+)
+
 // Client interface defines the clients need to implement to talk to CA for CSR.
 type Client interface {
-	CSRSign(csrPEM []byte /*PEM-encoded certificate request*/, subjectID string, certValidTTLInSec int64) ([]byte /*PEM-encoded certificate chain*/, error)
+	CSRSign(ctx context.Context, csrPEM []byte, /*PEM-encoded certificate request*/
```
Comment (wattli) [inline security/pkg/nodeagent/caClient/client.go]: Same here.

File: security/pkg/nodeagent/caClient/client.go [hunk 3]
```diff
@@ -14,7 +14,61 @@
...
+		log.Errorf("Failed to connect to CA: %v", err)
+		return nil, err
+	}
+
+	return &caClient{
+		client: capb.NewIstioCertificateServiceClient(conn),
+	}, nil
+}
+
+func (cl *caClient) CSRSign(ctx context.Context, csrPEM []byte, /*PEM-encoded certificate request*/
```
Comment (wattli) [inline security/pkg/nodeagent/caClient/client.go]: same here.

File: security/cmd/node_agent_k8s/main.go (hunk 1 above)
Comment (wattli) [reply security/cmd/node_agent_k8s/main.go]: and I think csrPEM itself is clear enough, btw :)

File: security/pkg/nodeagent/caClient/client.go [hunk 4]
```diff
@@ -14,7 +14,61 @@
...
+		return nil, err
+	}
+
+	if len(resp.CertChain) == 0 {
+		log.Errorf("Got empty cert chain")
+		return nil, errors.New("empty cert chain")
+	}
+
+	// Returns the leaf cert(Leaf cert is element '0', Root cert is element 'n').
+	return []byte(resp.CertChain[0]), nil
```
Comment (wattli) [inline security/pkg/nodeagent/caClient/client.go]: This should be a chain, not just the 0th one.

File: security/pkg/nodeagent/caClient/client_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+
+	capb "istio.io/istio/security/proto/ca/v1alpha1"
+)
+
+const (
+	mockServerAddress = "localhost:0"
```
Comment (wattli) [inline security/pkg/nodeagent/caClient/client_test.go]: use port other than 0?

File: security/pkg/nodeagent/caClient/client_test.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		}
+	}()
+
+	// The goroutine starting the server may not be ready, results in flakiness.
+	time.Sleep(1 * time.Second)
+
+	opts := []grpc.DialOption{grpc.WithInsecure()}
+	cli, err := NewCAClient(lis.Addr().String(), opts)
+	if err != nil {
+		t.Fatalf("failed to create ca client: %v", err)
```
Comment (wattli) [inline security/pkg/nodeagent/caClient/client_test.go]: Try to be consistent and use t.Errorf() only in the test? Errorf() won't stop the test and can show more errors, which makes the debug easier. Typically fatalF is only used in main function.

File: security/pkg/nodeagent/cache/secretcache_test.go [hunk 7]
```diff
@@ -44,7 +45,8 @@ func TestGetSecret(t *testing.T) {
 	}()
 
 	proxyID := "proxy1-id"
-	gotSecret, err := sc.GetSecret(proxyID, fakeSpiffeID, "jwtToken1" /*jwtToken*/)
+	ctx := context.Background()
+	gotSecret, err := sc.GetSecret(ctx, proxyID, fakeSpiffeID, "jwtToken1" /*jwtToken*/)
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache_test.go]: I think "/*jwtToken*/" is unnecessary here :)

File: security/pkg/nodeagent/caClient/client_test.go (hunk 5 above)
Comment (incfly) [reply security/pkg/nodeagent/caClient/client_test.go]: `0` makes it allocated from available ports.

File: security/cmd/node_agent_k8s/main.go (hunk 1 above)
Comment (quanjielin) [reply security/cmd/node_agent_k8s/main.go]: removed comment for this param(here and other places).

File: security/pkg/nodeagent/caClient/client.go (hunk 4 above)
Comment (quanjielin) [reply security/pkg/nodeagent/caClient/client.go]: updated.

File: security/pkg/nodeagent/cache/secretcache_test.go (hunk 7 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache_test.go]: removed.

File: security/pkg/nodeagent/caClient/client_test.go (hunk 6 above)
Comment (quanjielin) [reply security/pkg/nodeagent/caClient/client_test.go]: I think it's better to fail early(updated to use fatal in above errors like server failed to start,no need to continue in those cases),

File: security/pkg/nodeagent/caClient/client.go [hunk 8]
```diff
@@ -14,7 +14,66 @@
...
+	}
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("Authorization", fmt.Sprintf("Bearer %s", subjectID)))
+	resp, err := cl.client.CreateCertificate(ctx, req)
+	if err != nil {
+		log.Errorf("Failed to create certificate: %v", err)
+		return nil, err
+	}
+
+	if len(resp.CertChain) == 0 {
```
Comment (wattli) [inline security/pkg/nodeagent/caClient/client.go]: Should we check the cert chain size is larger than 1 as well?

File: security/pkg/nodeagent/cache/secretcache_test.go [hunk 9]
```diff
@@ -101,7 +103,7 @@ func TestRefreshSecret(t *testing.T) {
 		skipTokenExpireCheck = true
 	}()
 
-	_, err := sc.GetSecret("proxy1-id" /*proxyID*/, fakeSpiffeID, "jwtToken1" /*jwtToken*/)
+	_, err := sc.GetSecret(context.Background(), "proxy1-id" /*proxyID*/, fakeSpiffeID, "jwtToken1")
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache_test.go]: Same here for the proxyID

File: security/pkg/nodeagent/cache/secretcache_test.go [hunk 10]
```diff
@@ -134,7 +136,7 @@ func newMockCAClient() *mockCAClient {
 	return &cl
 }
 
-func (c *mockCAClient) CSRSign(csrPEM []byte, /*PEM-encoded certificate request*/
+func (c *mockCAClient) CSRSign(ctx context.Context, csrPEM []byte, /*PEM-encoded certificate request*/
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache_test.go]: Please remove the comment

File: security/pkg/nodeagent/caClient/client_test.go (hunk 5 above)
Comment (wattli) [reply security/pkg/nodeagent/caClient/client_test.go]: ok

File: security/pkg/nodeagent/cache/secretcache_test.go (hunk 9 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache_test.go]: removed

File: security/pkg/nodeagent/cache/secretcache_test.go (hunk 10 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache_test.go]: done

File: security/pkg/nodeagent/caClient/client.go (hunk 8 above)
Comment (quanjielin) [reply security/pkg/nodeagent/caClient/client.go]: updated

**Pull request merged**